Make Future store Storages instead of references to DataPtrs

lw opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#60470 Make Future store Storages instead of references to DataPtrs**

A Future needs to know what DataPtrs are used by its value, but it isn't always able to extract them (and even when it is, that's expensive) so they're cached. DataPtrs are kinda like unique_ptrs (movable only, cannot be copied) hence the Future can only hold _references_ to them. The Future's value, however, is unfortunately mutable (we'd wish that weren't the case, but we don't think we can prevent that), which means the tensor/storage that owned that DataPtr might be deleted and thus the DataPtr could be freed. This means our cached reference becomes stale! Which leads to all kinds of disaster, like reading garbage data or segfaulting.
Luckily all the DataPtrs we were dealing with were held inside Storages, which have a shared_ptr semantics, thus allowing us to hold a strong pointer to them which ensures they're kept alive.

Differential Revision: [D29303570](https://our.internmc.facebook.com/intern/diff/D29303570/)

lw committed:
Make Future store Storages instead of references to DataPtrs



Differential Revision: [D29303570](https://our.internmc.facebook.com/intern/diff/D29303570/)

[ghstack-poisoned]

mrshenli reviewed (approved):
Replacing reference to DataPtr to Storage LGTM. I have two comments for followups:

1. Shall we add a test that modifies Future's value after completion?
2. Would I be correct if it is possible to modify `c10/core/Storage.h` to expose a `weak_intrusive_ptr` from the `storage_impl_` so that we don't need to hold on to the storage when not necessary?

https://github.com/pytorch/pytorch/blob/44cc873fba5e5ffc4d4d4eef3bd370b653ce1ce1/c10/core/Storage.h#L162

lw committed:
Update on "Make Future store Storages instead of references to DataPtrs"


A Future needs to know what DataPtrs are used by its value, but it isn't always able to extract them (and even when it is, that's expensive) so they're cached. DataPtrs are kinda like unique_ptrs (movable only, cannot be copied) hence the Future can only hold _references_ to them. The Future's value, however, is unfortunately mutable (we'd wish that weren't the case, but we don't think we can prevent that), which means the tensor/storage that owned that DataPtr might be deleted and thus the DataPtr could be freed. This means our cached reference becomes stale! Which leads to all kinds of disaster, like reading garbage data or segfaulting.
Luckily all the DataPtrs we were dealing with were held inside Storages, which have a shared_ptr semantics, thus allowing us to hold a strong pointer to them which ensures they're kept alive.

Differential Revision: [D29303570](https://our.internmc.facebook.com/intern/diff/D29303570/)

[ghstack-poisoned]

mrshenli commented on torch/testing/_internal/distributed/rpc/rpc_test.py:
```diff
@@ -6180,3 +6180,34 @@ def test_async_execution_nested_with_cuda_future(self):
 
         rpc.shutdown()
+
+    @skip_if_lt_x_gpu(1)
```
tests LGTM. 

nit: since these tests do not use RPC, does it make sense to move them to https://github.com/pytorch/pytorch/blob/master/test/test_futures.py?

lw replied:
Ah! I didn't know we had such a file! But, yes, absolutely, that would be a much better home for these tests! I'll move them there in a follow-up PR, so that I can also move the previous Future tests without adding too much noise to this PR.
